### PR TITLE
Replace if_chain with let chains

### LINKED ISSUE #9353: Remove all `if_chain!` macros
### Description

Current Clippy code has a lot of legacy `if_chain!` macros even though the nightly [if let chains](https://rust-lang.github.io/rfcs/2497-if-let-chains.html) is already in use. I would like to remove all `if_chain!` macro usage, replacing it with the regular if statement, unless of course there is any reason to avoid that?

One thing to note is that `cargo fmt` still hasn't been decided (see [rfc](https://github.com/rust-dev-tools/fmt-rfcs/issues/169)) on how to handle it, and doesn't seem likely to be decided soon. This is not a blocker, as the code can always be reformatted later.

#### Possible Lint?
P.S. I just realized that this might be a good automated lint because it would be useful to other existing code as well. See #9354

### PR DESCRIPTION
**Pull request by Alexendoo**
Closes #9353

Let chains are now supported by rustfmt 🎉

The PR is split into two commits
1. The result of running [`if-to-let-chain clippy*/**/*.rs`](https://github.com/Alexendoo/if-to-let-chain)
2. The manual clean up: fixing some errors/formatting, dogfood lints, removing the if_chain internal lint

r? @flip1995

changelog: none

### CODE REVIEW

**Comment by flip1995** [comment]
I will review this today. No need to rebase before that. Once I'm done reviewing, you can rebase and directly merge :shipit:

File: clippy_lints/src/utils/internal_lints/if_chain_style.rs
Comment (flip1995) [inline clippy_lints/src/utils/internal_lints/if_chain_style.rs:1]: Non-blocking Question: IIRC this lint checked that infallible `let` bindings at the start or end of the `if_chain` macro were moved before or after the `if_chain` resp.? Would this same lint also make sense for `let-chains`?

**Review by flip1995 — APPROVED** [review]
LGTM overall. `rg if_chain` in the repo still shows a few `if_chain` usages. But those are in tests and necessary. However, I think once let-chains are stabilized, we can also remove this one special check for `if_chain` in the `index_refutable_slice` lint implementation.

The `needless_borrow_pat` test implements its own `if_chain` macro for testing. Maybe we can reuse that in `index_refutable_slice` for now and with that get rid of the `if_chain` dep in the top level `Cargo.toml`.

All of this is non-blocking for this PR. Feel free to address it here or we can deal with it in a follow-up PR.  r=me after rebase :shipit:

File: clippy_lints/src/utils/internal_lints/if_chain_style.rs
Comment (Jarcho) [reply clippy_lints/src/utils/internal_lints/if_chain_style.rs:1]: That's already caught by rustc as `irrefutable_let_patterns`.

**Comment by Alexendoo** [comment]
@bors r=flip1995

The `if_chain` implementation from `index_refutable_slice` didn't seem to trigger the lint so I've left the `if_chain` dev dep there for now

**Pull request merged**